### PR TITLE
Add index on short_backing_asset (Issue 960)

### LINKED ISSUE #960: Optimize iteration in asset_update_bitasset_evaluator
As described in https://github.com/bitshares/bitshares-core/pull/932#discussion_r190158358,
> For better performance, we can add a "by_backing_asset" index for `asset_bitasset_data_object` so can iterate through it here, rather then iterating through all bitAssets.

Related code is:
https://github.com/bitshares/bitshares-core/blob/0c75388457e0d6f999de89dc53514d5754a4f58c/libraries/chain/asset_evaluator.cpp#L380-L387

### PR DESCRIPTION
**Pull request by jmjatlanta**
This is a PR to fix #960 

A change was made for the recent hardfork that needs to iterate through all assets, looking for any asset that uses a certain asset as its backing asset.

This PR changes this linear iteration to an index search and a shorter iteration.

- [x] Add code to asset_evaluator to use this new index (was waiting for merge from hardfork to develop)
- [x] Move testing code from bitasset_tests2.cpp to bitasset_tests.cpp (also was waiting for the merge)

### CODE REVIEW

**Comment by ryanRfox** [comment]
@jmjatlanta May I request you move this onto the proper Project Board? It is not clear to me if it should be on a Protocol Upgrade or a Feature Release. I placed it in New on the Project Backlog for now. You may also remove the `0 Question` Label when finished. Thanks

**Comment by jmjatlanta** [comment]
Hehe, well @ryanRfox I think I did the right thing. I originally moved it into the feature release, but then realized this was a PR and not an issue. I then went to move the issue there, and it was already there.

So, issues belong on the project board, and not PRs, right? Let me know if my thinking is backward.

**Comment by ryanRfox** [comment]
TBH @jmjatlanta this whole process is still a work in progress and TBD. I'll provide my thoughts below about we may do it, but there is plenty of room for debate about how we should do it.

My feeling is that the underlying Issue should be developed and tested by the developer, Labeled `2 Approved` and placed in the Done column. However, I don't think it should be Closed yet. Next, the developer should make the PR, place that in the Testing column, invite your reviewers and Label it `2 Final Review`. The reviewer may kick the Issue back to the developer for refactoring and move it into the In Development column (PR stays In Testing). Ultimately the developer will have an Issue Labeled `2 Approved` and the reviewer will mark the PR the same. Once the PR is merged, it can be marked Closed and so should the Issue. Those steps may be able to be automated by GitHub. 

Again, I'm not here to dictate the process, rather learn how best we can use these tools to organize the Teams's standard process.

File: libraries/chain/asset_evaluator.cpp [hunk 1]
```diff
@@ -378,54 +378,60 @@ void check_children_of_bitasset(database& d, const asset_update_bitasset_operati
...
-   const auto& idx = d.get_index_type<asset_index>().indices().get<by_type>();
-
-   for( auto itr = idx.lower_bound(true); itr != idx.end(); ++itr )
+   const auto& idx = d.get_index_type<graphene::chain::asset_bitasset_data_index>().indices().get<by_short_backing_asset>();
+   auto backed_by_itr = idx.find( op.asset_to_update );
+   if ( backed_by_itr != idx.end() ) // if we found at least 1
    {
-      const auto& child = *itr;
-      if ( child.bitasset_data(d).options.short_backing_asset == op.asset_to_update )
+      auto backed_end = idx.upper_bound( new_backing_asset.get_id() );
```
Comment (abitmore) [inline libraries/chain/asset_evaluator.cpp]: This looks incorrect. Should be `idx.upper_bound( op.asset_to_update )`.

File: libraries/chain/asset_evaluator.cpp [hunk 2]
```diff
@@ -378,54 +378,60 @@ void check_children_of_bitasset(database& d, const asset_update_bitasset_operati
...
+
+      for( ; backed_by_itr != backed_end; ++backed_by_itr )
       {
-         if ( after_hf_922_931 )
+         const auto& bitasset_data = *backed_by_itr;
+         const auto& child = bitasset_data.asset_id(d);
+         if ( bitasset_data.options.short_backing_asset == op.asset_to_update )
```
Comment (abitmore) [inline libraries/chain/asset_evaluator.cpp]: After fixed the above one, this `if` is not needed.

File: libraries/chain/asset_evaluator.cpp (hunk 1 above)
Comment (jmjatlanta) [reply libraries/chain/asset_evaluator.cpp]: Ouch, nice find! Evidently tunnel vision set in on the lines above, and failed to follow through on this one.

File: libraries/chain/asset_evaluator.cpp (hunk 2 above)
Comment (jmjatlanta) [reply libraries/chain/asset_evaluator.cpp]: Avoiding this ``if`` was the purpose for this change. Thanks for being thorough and setting me straight.

File: libraries/chain/asset_evaluator.cpp [hunk 3]
```diff
@@ -378,13 +378,16 @@ void check_children_of_bitasset(database& d, const asset_update_bitasset_operati
       return;
 
    // loop through all assets that have this asset as a backing asset
-   const auto& idx = d.get_index_type<asset_index>().indices().get<by_type>();
-
-   for( auto itr = idx.lower_bound(true); itr != idx.end(); ++itr )
+   const auto& idx = d.get_index_type<graphene::chain::asset_bitasset_data_index>().indices().get<by_short_backing_asset>();
+   auto backed_by_itr = idx.find( op.asset_to_update );
+   if ( backed_by_itr != idx.end() ) // if we found at least 1
```
Comment (pmconrad) [inline libraries/chain/asset_evaluator.cpp]: I think this if is not needed.

File: libraries/chain/asset_evaluator.cpp [hunk 4]
```diff
@@ -424,8 +427,8 @@ void check_children_of_bitasset(database& d, const asset_update_bitasset_operati
                }
             } // if child.issuer
          } // if hf 922/931
-      } // if this child is backed by the asset being adjusted
-   } // for each asset
+      } // for each asset
```
Comment (pmconrad) [inline libraries/chain/asset_evaluator.cpp]: ...backed by asset_to_update

File: libraries/chain/asset_evaluator.cpp (hunk 3 above)
Comment (jmjatlanta) [reply libraries/chain/asset_evaluator.cpp]: Unfortunately it is needed. In index [ 2, 4, 6, 8 ] ``upper_bound(5)`` points to 6, ``find(5)`` points to 8, so ``upper_bound != find``, and the ``for`` below enters when it is not supposed to, and possibly (gulp!) never ends.

File: libraries/chain/asset_evaluator.cpp (hunk 4 above)
Comment (jmjatlanta) [reply libraries/chain/asset_evaluator.cpp]: Done

File: libraries/chain/asset_evaluator.cpp (hunk 3 above)
Comment (pmconrad) [reply libraries/chain/asset_evaluator.cpp]: How about using https://www.boost.org/doc/libs/1_55_0/libs/multi_index/doc/reference/ord_indices.html#range_operations instead?

Comment (jmjatlanta) [reply libraries/chain/asset_evaluator.cpp]: Switched to using range. Take a look. The lambda syntax there looks hard to read IMO, but maybe it is just me. Had to include an extra boost header that throws deprecation warnings, but it is necessary for boost <= 1.63.

I am unaware of find() not working as expected on non-unique indexes. If you could point me to anything on the subject I'd appreciate it. If nothing is handy, I'll just keep it in mind until I run across something that proves or disproves it.

Comment (abitmore) [reply libraries/chain/asset_evaluator.cpp]: I'd rather code like this
```
auto itr = idx.lower_bound( xxx );
auto end = idx.upper_bound( xxx );
while( itr != end )
{
   process();
   ++itr;
}
```

If using `range` something (although I don't like lambda)
```
auto range = idx.equal_range( xxx );
std::for_each(range.first, range.second,
                    [&something](const yyy_obj& zzz) {
                       process();
                    });
```

Comment (jmjatlanta) [reply libraries/chain/asset_evaluator.cpp]: I think that looks a lot cleaner. Thanks @abitmore and @pmconrad. Testing now, should be checked in soon.

File: libraries/chain/asset_evaluator.cpp [hunk 5]
```diff
@@ -378,54 +378,55 @@ void check_children_of_bitasset(database& d, const asset_update_bitasset_operati
...
-      const auto& child = *itr;
-      if ( child.bitasset_data(d).options.short_backing_asset == op.asset_to_update )
-      {
-         if ( after_hf_922_931 )
+   const auto& idx = d.get_index_type<graphene::chain::asset_bitasset_data_index>()
+         .indices()
+         .get<by_short_backing_asset>();
+   auto backed_range = idx.equal_range(op.asset_to_update);
+   std::for_each( backed_range.first, backed_range.second,
+         [after_hf_922_931, &new_backing_asset, &d, &op](const asset_bitasset_data_object& bitasset_data)
```
Comment (abitmore) [inline libraries/chain/asset_evaluator.cpp]: I guess if change this indentation to smaller, the whole diff will look much better.

Comment (pmconrad) [reply libraries/chain/asset_evaluator.cpp]: Hint: diff -w (or "hide whitespace" in diff settings on github)

Comment (abitmore) [reply libraries/chain/asset_evaluator.cpp]: Thanks. That does the trick.

File: libraries/chain/asset_evaluator.cpp (hunk 3 above)
Comment (pmconrad) [reply libraries/chain/asset_evaluator.cpp]: > I am unaware of find() not working as expected on non-unique indexes.

An ordered non-unique boost multi-index is equivalent to a std::multiset. Look at the definition of std::multiset::find - it "Finds an element with key equivalent to key", but doesn't make any other guarantees.

**Pull request merged**